grpc/server.go

```diff
@@ -0,0 +1,288 @@
+ * identify the disruption to apply to a query. The mapping represents user preference the proportion of queries
+ * affected by each alteration. See ALGORITHM.md for examples.
+ */
+
+// DisruptionListener is a gRPC Service that can disrupt endpoints of a gRPC server.
+type DisruptionListener struct {
+	pb.UnimplementedDisruptionListenerServer
+	Configuration DisruptionConfiguration
+}
+
+// DisruptionConfiguration configures the DisruptionListener to chaos test endpoints of a gRPC server.
+type DisruptionConfiguration map[TargetEndpoint]EndpointConfiguration
+
+// EndpointConfiguration configures endpoints that the DisruptionListener chaos tests on a gRPC server.
+// The AlterationHash maps integers from 0 to 100 to alteration configurations.
+type EndpointConfiguration struct {
+	TargetEndpoint TargetEndpoint
+	Alterations    map[AlterationConfiguration]PercentAffected
+	AlterationHash map[PercentSlot]AlterationConfiguration
+}
+
+// AlterationConfiguration contains either an ErrorToReturn or an OverrideToReturn for a given
+// gRPC query to the disrupted service.
+type AlterationConfiguration struct {
+	ErrorToReturn    string
+	OverrideToReturn string
+}
+
+// PercentAffected is an integer represented the number of queries out of 100 queries which should
+// be affected by a certain endpoint alteration.
+type PercentAffected int
+
+// PercentSlot is an integer used to make a random, weighted assignment to an endpoint based on the
+// configured PercentAffected per endpoint in a GrpcDisruptionConfiguration.
+type PercentSlot int
+
+// TargetEndpoint is a string of the format /package.service/method.
+type TargetEndpoint string
+
+var (
+	log *zap.SugaredLogger
+
+	errorMap = map[string]codes.Code{
+		"OK":                  codes.OK,
+		"CANCELED":            codes.Canceled,
+		"UNKNOWN":             codes.Unknown,
+		"INVALID_ARGUMENT":    codes.InvalidArgument,
+		"DEADLINE_EXCEEDED":   codes.DeadlineExceeded,
+		"NOT_FOUND":           codes.NotFound,
+		"ALREADY_EXISTS":      codes.AlreadyExists,
+		"PERMISSION_DENIED":   codes.PermissionDenied,
+		"RESOURCE_EXHAUSTED":  codes.ResourceExhausted,
+		"FAILED_PRECONDITION": codes.FailedPrecondition,
+		"ABORTED":             codes.Aborted,
+		"OUT_OF_RANGE":        codes.OutOfRange,
+		"UNIMPLEMENTED":       codes.Unimplemented,
+		"INTERNAL":            codes.Internal,
+		"UNAVAILABLE":         codes.Unavailable,
+		"DATA_LOSS":           codes.DataLoss,
+		"UNAUTHENTICATED":     codes.Unauthenticated,
+	}
+
+	mutex sync.Mutex
+)
+
+// SendDisruption makes a gRPC call to DisruptionListener.
+func (d *DisruptionListener) SendDisruption(ctx context.Context, ds *pb.DisruptionSpec) (*emptypb.Empty, error) {
+	if ds == nil {
+		log.Error("Cannot execute SendDisruption when DisruptionSpec is nil")
+		return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption when DisruptionSpec is nil")
+	}
+
+	config := DisruptionConfiguration{}
+
+	for _, endpointSpec := range ds.Endpoints {
```

Comment:
Looking into it some more, there's a 3rd structure (the pb item). However, i think it's necessary to transfer the data over (and I can't reuse it with kubebuilder because that's tied to the the CRD. The final format with the percentages is also necessary with the current design. Let me know if you disagree I can wrestle with it some more!

Reply:
Hmm, does seem like I was wrong then, though I do wish there were some way to repeat this validation less often.